modules/flann/src/miniflann.cpp

```diff
@@ -482,15 +482,18 @@ void runKnnSearch_(void* index, const Mat& query, Mat& indices, Mat& dists,
     typedef typename Distance::ResultType DistanceType;
     int type = DataType<ElementType>::type;
     int dtype = DataType<DistanceType>::type;
+    IndexType* index_ = (IndexType*)index;
+
+    CV_Assert(knn <= index_->size());
```

Comment:
There are two main strategies:
- Failure-avoidance strategies
    - Set `knn = MIN(knn, index_->size());`
    - Swallow it and return NaNs for the additional kNNs.
- Raise and exception to reject the user input

Is it my imagination, or this is also related to #8300? In that case, I guess the vote for this matter is meaningless, as there are cases where one strategy is better than the other, and vise versa.

In this specific case, the user is expecting a matrix of kNN width, returning smaller one is not a sane choice.
Moreover, while returning NaNs in the distance makes sense, the indices matrix needs to be `int`-type. Accordingly, while I am a fan of failure-avoidance strategies, it just does not make sense here, and an exception is the best choice.